### PR TITLE
Accept valid coordinates as input

### PR DESCRIPTION
**Pull request by thorncp**
There're some refactors lurking here that I want to hold off for follow-up PRs. I'm thinking the next step, using the input to redraw the grid, will be a good time for that.

### CODE REVIEW

File: spec/game_spec.rb [hunk 1]
```diff
@@ -2,12 +2,14 @@
...
+    stdin = StringIO.new
     stdout = StringIO.new
-    game = Game.new(output: stdout)
+    game = Game.new(input: stdin, output: stdout)
+    allow(stdin).to receive(:gets).and_return("R2", "3PO", "C1")
 
     game.play
 
-    expect(stdout.string).to include <<~GRID
+    expect(stdout.string).to eq(<<~GRID.chomp)
```
Comment (baccigalupi) [inline spec/game_spec.rb]: This is a great test!

File: lib/game.rb [hunk 2]
```diff
@@ -8,17 +8,48 @@ class Game
       |__|__|__|
     C |  |  |  |
       |__|__|__|
+
   GRID
```
Comment (baccigalupi) [inline lib/game.rb]: To me all these constants are starting to look very messy and distract from the logic. And in terms of color and shape they seem to hold together as a printer class. Extracting those from the game play will allow this looping runner to be more clear.

File: lib/game.rb [hunk 3]
```diff
@@ -8,17 +8,48 @@ class Game
...
     output.puts(CLI_GRID)
+    until coordinates
+    end
   end
 
   private
 
-  attr_reader :output
+  attr_reader :input, :output
+
```
Comment (baccigalupi) [inline lib/game.rb:19]: There are a bunch of methods that mix concerns here ...

File: lib/game.rb [hunk 4]
```diff
@@ -8,17 +8,48 @@ class Game
...
+  def coordinates
```
Comment (baccigalupi) [inline lib/game.rb]: I would expect `coordinates` to return something, but not do something. This is mixing the query and command in a way that isn't intuitive.

File: lib/game.rb [hunk 5]
```diff
@@ -8,17 +8,48 @@ class Game
...
+    validate(prompt)
+  end
+
+  def prompt
```
Comment (baccigalupi) [inline lib/game.rb]: Here too, there is output and then input, which seem like very different concerns. Also I wouldn't expect the prompt to get input. I would just expect it to print out the prompt

File: lib/game.rb [hunk 6]
```diff
@@ -8,17 +8,48 @@ class Game
...
+    output.write(CLI_MOVE_PROMPT)
+    input.gets.chomp
+  end
+
+  def validate(entered_coordinates)
```
Comment (baccigalupi) [inline lib/game.rb]: This too is mixing returning a value with printing something out. The returned value is a query type method. And the printing is a command. So query/command separation would be good.

File: spec/game_spec.rb [hunk 7]
```diff
@@ -16,6 +18,14 @@
...
+
+      Enter your move >
+      Invalid coordinates 'R2'
+
+      Enter your move >
+      Invalid coordinates '3PO'
+
+      Enter your move >
     GRID
   end
```
Comment (baccigalupi) [inline spec/game_spec.rb:29]: I would expect more tests that are unit-y and focused around boundary conditions for coordinates. Like a coordinate is off the grid in any direction. Those kind of unit tests mean that you would have a separate object you could unit test.

**Review by baccigalupi — CHANGES_REQUESTED** [review]
I would typically approve a PR like this even with concerns because it is pretty small, and as the chaos mounts the patterns are ofter easier to see. But since this is an exercise, I am sending you back to this PR for more work.

**Comment by thorncp** [comment]
Extracted a couple classes. I'm struggling a bit with modeling the objects to get user input. The looping validation bit feels unlike what I'm used to and I resorted to keeping some state that feels _wrong_, but I'm not sure how to move forward with it. Any guidance?

File: lib/game.rb [hunk 8]
```diff
@@ -1,24 +1,30 @@
...
+require_relative "player_input"
 
-  def initialize(output:)
+class Game
+  def initialize(input:, output:)
+    @input = input
     @output = output
   end
 
   def play
```
Comment (baccigalupi) [inline lib/game.rb:11]: This is very readable!

File: lib/player_input.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  def get_and_validate_input
+    print_prompt
+    get_input
+    unless validate_input
+      print_error
+    end
+  end
+
+  def get_input
+    @entered_coordinates = input.gets.chomp
```
Comment (thorncp) [inline lib/player_input.rb]: I've found it challenging to model the loop of getting and validating input. Capturing some state here and in the method below feels _wrong_, but I can't put into words why. I tried pulling out another class, but that felt worse somehow. I feel like there's something I'm missing or I'm overcomplicating things. Any guidance?

File: lib/grid_view.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+class GridView
```
Comment (baccigalupi) [inline lib/grid_view.rb:1]: Love the names. In fact when I saw your PR I changed the names of stdin and stdout in my program. But all the method names are clear and make sense.

File: lib/player_input.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
+class PlayerInput
```
Comment (baccigalupi) [inline lib/player_input.rb]: This class is less intuitive to me, and that comes across with the size and complexity of the class and methods. That it needs to output and handle input seems like too much responsibility.

What about something that just wraps input. This might be two classes. One that coordinates a turn, printing the prompt and then getting/validating input, and one that wraps the input?

File: lib/player_input.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  def get_input
+    @entered_coordinates = input.gets.chomp
+  end
+
+  def validate_input
+    if valid_coordinates.include?(entered_coordinates)
+      @validated_coordinates = entered_coordinates
+    end
+  end
+
```
Comment (baccigalupi) [inline lib/player_input.rb]: These printing related methods hold together on their own, but not with the rest of the methods. I see that because of the names of the methods and also their main concern being output.

File: spec/player_input_spec.rb [hunk 13]
```diff
@@ -0,0 +1,35 @@
+require_relative "../lib/player_input"
+
+RSpec.describe PlayerInput do
```
Comment (baccigalupi) [inline spec/player_input_spec.rb]: These tests still seem pretty integration focused. We need to inspect the output in order to test whether the coordinates are validated. Earlier I was talking about a possible `Turn` class and a wrapper class for input, but it might be that there is a data validation class hidden in there.

**Review by baccigalupi — COMMENTED** [review]
The PlayerInput class feels like it is wearing too many hats in this one. But I loved a lot of the other cleanups and in particular naming.

File: lib/player_input.rb (hunk 12 above)
Comment (thorncp) [reply lib/player_input.rb]: Extracted another view object in d733979. I've started to feel a pull to having these view objects be injected, instead of the output stream itself, but I'm not sure I'm ready to make that jump just yet.

File: lib/player_input.rb (hunk 11 above)
Comment (thorncp) [reply lib/player_input.rb]: Pulled out a few things in 2f3dcb7 and d733979. I like the "turn" mentioned below, so renamed this class to PlayerTurn in 875d678.

File: spec/player_input_spec.rb (hunk 13 above)
Comment (thorncp) [reply spec/player_input_spec.rb]: Updated to inject a view object instead of the output stream in 5d2a0380db819c28395dc3e2766e56b18889b748. It does feel much less integration focused, but I don't really like the heavy use of mocking - I don't feel super confident the test has me covered. I'm gonna keep poking at it, but figured I'd share what's there for now.

File: lib/coordinates.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  def self.valid_values
+    %w[A1 A2 A3 B1 B2 B3 C1 C2 C3]
+  end
+
+  def initialize(value:)
+    @value = value
+  end
+
+  def valid?
+    self.class.valid_values.include?(value)
```
Comment (baccigalupi) [inline lib/coordinates.rb]: I would typically make `self.class.valid_values` into an instance method that mirrors the class method. That would make this a lot easier to read: `valid_values.include?(value)`

File: lib/game.rb [hunk 15]
```diff
@@ -1,24 +1,35 @@
...
 
-  attr_reader :output
+  attr_reader :input, :output
+
+  def get_player_input
+    player_input.get_coordinates
+  end
+
+  def player_input
+    PlayerTurn.new(input: input, view: text_view)
```
Comment (baccigalupi) [inline lib/game.rb:25]: Oooh, having a PlayerTurn class! Excited to see it.

File: lib/game.rb [hunk 16]
```diff
@@ -1,24 +1,35 @@
...
+  end
+
```
Comment (baccigalupi) [inline lib/game.rb:27]: Nice! Seeing these view objects is cool. Also, when I see a top level coordinator class coalescing into these object instantiations and then DI, it feels good.

File: lib/null_coordinates.rb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+class NullCoordinates
```
Comment (baccigalupi) [inline lib/null_coordinates.rb:1]: OMG! Love Null objects. Thank you!

File: lib/player_turn.rb [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+class PlayerTurn
+  attr_reader :input, :view
+
+  def initialize(input:, view:)
+    @input = input
+    @view = view
+  end
+
+  def get_coordinates
+    until coordinates.valid?
```
Comment (baccigalupi) [inline lib/player_turn.rb:13]: This is so sweet! I love how the null object cleared up so much logic.

File: lib/player_turn.rb [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    coordinates.value
+  end
+
+  private
+
+  def coordinates
+    @coordinates || NullCoordinates.new
+  end
+
+  def get_and_validate_coordinates
```
Comment (baccigalupi) [inline lib/player_turn.rb:25]: Ahh! This is understandably complex, but the naming and responsibilities are lining up to makes for better comprehension.

File: lib/text_view.rb [hunk 20]
```diff
@@ -0,0 +1,19 @@
+class TextView
```
Comment (baccigalupi) [inline lib/text_view.rb:1]: Love it!

File: spec/coordinates_spec.rb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+    it "is true when the value is in the grid" do
+      described_class.valid_values.each do |value|
+        coordinates = described_class.new(value: value)
+
+        expect(coordinates).to be_valid
+      end
+    end
+
+    it "is false when the value is not in the grid" do
+      boundaries = %w[A0 B0 C0 A4 B4 C4 D1 D2 D3]
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb]: This weekend I was reading a blog post by Justin Searls at Test Double about the right level and amount of testing. It's really good. Reading it could make these specs better. 

I am not a fan of putting a loop into a spec. It is pretty hard to get your heard around. In this case the data that generates the loop also requires a bit too much knowledge. 

I think there are real concerns in these tests that don't need to as exhaustive:
* When in a row before the first row of the grid
* When in a row after the last row in the grid
* When in a column before the first column in the grid
* When in a column after the last column of the grid

The other specs are very exhaustive, but not expressing the most important logic.

File: spec/coordinates_spec.rb [hunk 22]
```diff
@@ -0,0 +1,31 @@
+require_relative "../lib/coordinates"
+
+RSpec.describe Coordinates do
+  describe "#valid?" do
+    it "is true when the value is in the grid" do
+      described_class.valid_values.each do |value|
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb]: I left a comment below about loops in tests and exhaustiveness, and looped back up to this spec. I don't think we need to test every valid coordinate. Also exposing the use of this class method/data feels bad. Typically I think of anything not tested as private, and this data feels like it doesn't belong in a test.

I think testing boundaries with a spec each makes more sense than testing every coordinate.

Comment (baccigalupi) [reply spec/coordinates_spec.rb]: One more thing about this spec that shows up in other places. I have found `subject` and also `described_class` to be much less clear than using the class constant itself. 

I like my tests to be completely unclever and duplicative so that it is clear what is happening.

File: spec/player_turn_spec.rb [hunk 23]
```diff
@@ -0,0 +1,33 @@
+require_relative "../lib/player_turn"
+
+RSpec.describe PlayerTurn do
+  describe "#get_coordinates" do
```
Comment (baccigalupi) [inline spec/player_turn_spec.rb:4]: Nice! There is some complex logic in this universe of taking a turn, and having these collaborators with their own responsibilities allows these tests to setup up exact situations and expectations!

**Review by baccigalupi — CHANGES_REQUESTED** [review]
I have some strong opinions about that one spec file, but the rest looks great!

File: spec/coordinates_spec.rb (hunk 21 above)
Comment (thorncp) [reply spec/coordinates_spec.rb]: I like that reasoning. Updating in 9c06d7236f0f90754186c91317fadaa306dc725.

I'm not sure what a good test value for before the first row is, though. What comes before A?

File: lib/player_turn.rb (hunk 18 above)
Comment (thorncp) [reply lib/player_turn.rb:13]: Yeah, that part was enjoyable.

File: spec/coordinates_spec.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
+require_relative "../lib/coordinates"
+
+RSpec.describe Coordinates do
+  describe "#valid?" do
+    it "is true when the value is in the grid" do
+      expect(Coordinates.new(value: "A1")).to be_valid
+      expect(Coordinates.new(value: "B2")).to be_valid
+      expect(Coordinates.new(value: "C3")).to be_valid
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb:8]: Nice. That is a great way to exercise the boundaries.

File: spec/coordinates_spec.rb [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
+    it "is false when the value is not in the grid" do
+      expect(Coordinates.new(value: "D1")).not_to be_valid
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb]: I would call out what about each one of these examples is not acceptable. So we don't have to know about the game coordinates to understand at a glance. So I would recommend that each example be its own spec, with a good description of why it is invalid.

File: spec/null_coordinates_spec.rb [hunk 26]
```diff
@@ -0,0 +1,21 @@
+require_relative "../lib/null_coordinates"
+
+RSpec.describe NullCoordinates do
+  describe "#value" do
+    it "is nil" do
+      expect(described_class.new.value).to be_nil
```
Comment (baccigalupi) [inline spec/null_coordinates_spec.rb]: I'd recommend swapping out these `described_class` calls too.

**Review by baccigalupi — APPROVED** [review]
Ton better spec-ing!

File: spec/coordinates_spec.rb (hunk 25 above)
Comment (thorncp) [reply spec/coordinates_spec.rb]: Oh, that's a good idea.

Comment (thorncp) [reply spec/coordinates_spec.rb]: Updated in f97bc00fc59611a7fc8cd418987d6aec5c6e738e

**Comment by thorncp** [comment]
Thanks for the thorough reviews, @baccigalupi !
